### PR TITLE
Add mbart

### PR DESCRIPTION
**Pull request by Zhylkaaa**
This PR intends to add mBART model and related to https://github.com/huggingface/transformers/issues/5753. 
previously u can't use `facebook/mbart-large-cc25` or any mbart models in seq2seq.
It seems to fix problem for me :)

### CODE REVIEW

**Comment by ThilinaRajapakse** [comment]
Thank you!

But, it looks like the tokenization needs to be handled differently for MBart models [(HF docs](https://huggingface.co/transformers/model_doc/mbart.html#mbarttokenizer)). `Seq2SeqModel` will use `tokenizer.encode()` which seems to be incompatible with MBart.

**Comment by Zhylkaaa** [comment]
sorry, hasn't noticed this change but I am not sure it was present in v3.3.1 docs. But actually I can't see tokenizer return `decoder_input_ids` even though I am passing `tgt_texts`, try to do this and see, maybe it's tokenizers problem on my machine 
```
from transformers import MBartForConditionalGeneration, MBartTokenizer
tokenizer = MBartTokenizer.from_pretrained('facebook/mbart-large-cc25')
tokenizer.prepare_seq2seq_batch(src_texts=['hi my nam is dima'], tgt_texts=['name dima']).keys()
```

Also I wanted to ask why don't you use `labels` and `decoder_attention_mask` for bart? Or I can add this too?

**Comment by ThilinaRajapakse** [comment]
It looks like you need to pass `src_lang` and `tgt_lang` as well.

```python
from transformers import MBartTokenizer
tokenizer = MBartTokenizer.from_pretrained('facebook/mbart-large-en-ro')
example_english_phrase = " UN Chief Says There Is No Military Solution in Syria"
expected_translation_romanian = "Şeful ONU declară că nu există o soluţie militară în Siria"
batch: dict = tokenizer.prepare_seq2seq_batch(
    example_english_phrase, src_lang="en_XX", tgt_lang="ro_RO", tgt_texts=expected_translation_romanian
)
```
So, `src_lang` and `tgt_lang` probably should be added to `Seq2SeqArgs` and then the tokenizer can grab it from there.

Essentially, the preprocess function for `mbart` should probably look like below (following the `preprocess_data_bart()` function).

```python
def preprocess_data_bart(data):
    input_text, target_text, tokenizer, args = data

    input_ids = tokenizer.prepare_seq2seq_batch(
        [input_text], src_lang=args["src_lang"], tgt_lang=args["tgt_lang"], tgt_texts=[target_text], max_length=args.max_seq_length, padding="max_length", return_tensors="pt", truncation=True
    )

    target_ids = tokenizer.batch_encode_plus(
        [target_text], max_length=args.max_seq_length, padding="max_length", return_tensors="pt", truncation=True
    )

    return {
        "source_ids": input_ids["input_ids"].squeeze(),
        "source_mask": input_ids["attention_mask"].squeeze(),
        "target_ids": target_ids["input_ids"].squeeze(),
    }
```

**Comment by ThilinaRajapakse** [comment]
The `decoder_attention_mask` is computed automatically from `decoder_input_ids`. The `labels` argument used to be `lm_labels` back when Bart was added to Simple Transformers, but yeah, it should be renamed to `labels`. (This is all done in `_get_inputs_dict()` method in `Seq2SeqModel`)

**Comment by Zhylkaaa** [comment]
the reason I created labels and decoder_ids directly in `preprocess_data_mbart` is that I think labels should have additional modification because `shift_tokens_right` actually leaves tgt_lang token at the and of sequence (before pad), and I think this shouldn't be taken into account in loss function. Or should I leave it as it is?

**Review by ThilinaRajapakse — APPROVED** [review]
I think this looks good.

Amazing work btw, thank you!

**Comment by parthplc** [comment]
Hey @ThilinaRajapakse,  So now MBART is supported for fine-tuning using SimpleTransformers using seq2seq script provided in docs?

**Comment by parthplc** [comment]
I am using this basic starting script
```
# logging.basicConfig(level=logging.INFO)
# transformers_logger = logging.getLogger("transformers")
# transformers_logger.setLevel(logging.WARNING)
from simpletransformers.seq2seq.seq2seq_model import Seq2SeqModel

train_data =  [
    ["एक", "1"],
    ["दो", "2"],
]

train_df = pd.DataFrame(train_data, columns=["input_text", "target_text"])

eval_data =[
    ["तीन", "3"],
    ["चार", "4"],
]

eval_df = pd.DataFrame(eval_data, columns=["input_text", "target_text"])

model_args = {
    "reprocess_input_data": True,
    "overwrite_output_dir": True,
    "max_seq_length": 10,
    "train_batch_size": 1,
    "num_train_epochs": 20,
    "save_eval_checkpoints": False,
    "save_model_every_epoch": False,
    "evaluate_during_training": True,
    "evaluate_generated_text": True,
    "evaluate_during_training_verbose": True,
    "use_multiprocessing": False,
    "max_length": 15,
    "manual_seed": 4,
    "src_lang": "hi_IN",
    "tgt_lang": "hi_IN"

}

# Initialize model
model = Seq2SeqModel(
    encoder_decoder_type="mbart",
    encoder_decoder_name="facebook/mbart-large-cc25",
    args=model_args,
)

# Train the model
model.train_model(train_df,eval_data=eval_df)

# Evaluate the model
results = model.eval_model(eval_df)

# Use the model for prediction
print(model.predict(["पाँच"]))
```

**Comment by Zhylkaaa** [comment]
Hi @parthplc, for me it looks fine. Note that you have 2 Hindi languages: one you are using and Gujarati (gu_IN)
you can see all languages available using:
```
from transformers import MBartTokenizer
tokenizer = MBartTokenizer.from_pretrained('facebook/mbart-large-cc25')
tokenizer.additional_special_tokens
```
but I think you are using right `src_lang` and `tgt_lang`
Are you facing any problems?

**Comment by parthplc** [comment]
Hey @Zhylkaaa, The above script runs fine. Since I will be finetuning on much larger Hindi dataset just wanted to ensure that this is the right ways. Thanks!
